File: server/routes/chatbot.js
/**
 * Chatbot API Routes
 */

import express from 'express'
import ChatSession from '../models/ChatSession.js'
import ChatMessage from '../models/ChatMessage.js'
import { protect } from '../middleware/auth.js'
import { detectIntent, generateResponse, createTicketFromChat } from '../services/chatbot.js'
import upload from '../middleware/upload.js'

const router = express.Router()

/**
 * @route   POST /api/chatbot/session
 * @desc    Create or get chat session
 * @access  Private
 */
router.post('/session', protect, async (req, res) => {
  try {
    // Find active session or create new one
    let session = await ChatSession.findOne({
      user: req.user._id,
      status: { $in: ['active', 'escalated'] },
    })
      .populate('assignedTo', 'name email')
      .populate('department', 'name')

    if (!session) {
      session = await ChatSession.create({
        user: req.user._id,
        organization: req.user.organization,
        metadata: {
          userAgent: req.headers['user-agent'],
          ipAddress: req.ip,
          platform: req.body.platform || 'web',
        },
      })
    }

    // Get messages
    const messages = await ChatMessage.find({ session: session._id })
      .populate('senderId', 'name email')
      .sort({ createdAt: 1 })

    res.json({
      session: {
        _id: session._id,
        sessionId: session.sessionId,
        status: session.status,
        assignedTo: session.assignedTo,
        department: session.department,
        ticketId: session.ticketId,
        createdAt: session.createdAt,
      },
      messages,
    })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/chatbot/message
 * @desc    Send message to chatbot
 * @access  Private
 */
router.post('/message', protect, upload.array('attachments', 5), async (req, res) => {
  try {
    const { message, sessionId } = req.body

    if (!message && (!req.files || req.files.length === 0)) {
      return res.status(400).json({ message: 'Message or attachment is required' })
    }

    // Get or create session
    let session = await ChatSession.findOne({
      $or: [
        { sessionId },
        { user: req.user._id, status: { $in: ['active', 'escalated'] } },
      ],
    })

    if (!session) {
      session = await ChatSession.create({
        user: req.user._id,
        organization: req.user.organization,
        metadata: {
          userAgent: req.headers['user-agent'],
          ipAddress: req.ip,
          platform: req.body.platform || 'web',
        },
      })
    }

    // Handle attachments
    const attachments = (req.files || []).map(file => ({
      filename: file.filename,
      path: file.path,
      mimetype: file.mimetype,
      size: file.size,
    }))

    // Create user message
    const userMessage = await ChatMessage.create({
      session: session._id,
      sender: 'user',
      senderId: req.user._id,
      content: message || 'File attachment',
      messageType: attachments.length > 0 ? 'file' : 'text',
      attachments,
    })

    // Detect intent
    const intent = await detectIntent(message || 'file', req.user._id, req.user.organization)

    // Generate bot response
    const botResponse = await generateResponse(intent, session, message || 'file', req.user._id)

    // Create bot message
    const botMessage = await ChatMessage.create({
      session: session._id,
      sender: 'bot',
      content: botResponse.content,
      messageType: botResponse.messageType || 'text',
      intent: intent.intent,
      confidence: intent.confidence,
      metadata: botResponse.metadata || {},
    })

    // Update session
    session.messages.push(userMessage._id, botMessage._id)
    await session.save()

    res.json({
      userMessage,
      botMessage: {
        ...botMessage.toObject(),
        quickActions: botResponse.quickActions,
        faqId: botResponse.faqId,
      },
    })
  } catch (error) {
    console.error('Chatbot message error:', error)
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/chatbot/create-ticket
 * @desc    Create ticket from chat
 * @access  Private
 */
router.post('/create-ticket', protect, async (req, res) => {
  try {
    const { sessionId, title, description, priority, category, department } = req.body

    if (!title || !description) {
      return res.status(400).json({ message: 'Title and description are required' })
    }

    // Get session
    const session = await ChatSession.findOne({
      $or: [
        { sessionId },
        { user: req.user._id, status: { $in: ['active', 'escalated'] } },
      ],
    })

    if (!session) {
      return res.status(404).json({ message: 'Chat session not found' })
    }

    // Create ticket
    const ticket = await createTicketFromChat(session, {
      title,
      description,
      priority: priority || 'medium',
      category: category || 'General',
      department,
    }, req.user._id)

    // Create system message
    const systemMessage = await ChatMessage.create({
      session: session._id,
      sender: 'bot',
      content: `Ticket #${ticket.ticketId} has been created successfully! You can track its status here or ask me anytime.`,
      messageType: 'ticket_created',
      metadata: { ticketId: ticket.ticketId },
    })

    session.messages.push(systemMessage._id)
    await session.save()

    res.json({
      ticket,
      message: systemMessage,
    })
  } catch (error) {
    console.error('Create ticket from chat error:', error)
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   GET /api/chatbot/history
 * @desc    Get chat history (for admins/technicians)
 * @access  Private
 */
router.get('/history', protect, async (req, res) => {
  try {
    const { userId, sessionId, limit = 50 } = req.query
    const query = {}

    // Regular users can only see their own chats
    if (req.user.role === 'user') {
      query.user = req.user._id
    } else if (userId) {
      query.user = userId
    }

    if (sessionId) {
      query.sessionId = sessionId
    }

    const sessions = await ChatSession.find(query)
      .populate('user', 'name email')
      .populate('assignedTo', 'name email')
      .populate('department', 'name')
      .sort({ createdAt: -1 })
      .limit(parseInt(limit))

    res.json(sessions)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   GET /api/chatbot/session/:sessionId
 * @desc    Get chat session details
 * @access  Private
 */
router.get('/session/:sessionId', protect, async (req, res) => {
  try {
    const session = await ChatSession.findOne({ sessionId: req.params.sessionId })
      .populate('user', 'name email')
      .populate('assignedTo', 'name email')
      .populate('department', 'name')
      .populate('ticket', 'ticketId title status')

    if (!session) {
      return res.status(404).json({ message: 'Session not found' })
    }

    // Check access
    if (req.user.role === 'user' && session.user._id.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Access denied' })
    }

    const messages = await ChatMessage.find({ session: session._id })
      .populate('senderId', 'name email')
      .sort({ createdAt: 1 })

    res.json({
      session,
      messages,
    })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/chatbot/escalate
 * @desc    Escalate chat to technician
 * @access  Private
 */
router.post('/escalate', protect, async (req, res) => {
  try {
    const { sessionId, departmentId } = req.body

    const session = await ChatSession.findOne({
      $or: [
        { sessionId },
        { user: req.user._id, status: 'active' },
      ],
    })

    if (!session) {
      return res.status(404).json({ message: 'Session not found' })
    }

    session.status = 'escalated'
    session.escalatedAt = new Date()

    if (departmentId) {
      session.department = departmentId
    }

    // Find available technician
    const Department = (await import('../models/Department.js')).default
    const User = (await import('../models/User.js')).default

    if (departmentId) {
      const department = await Department.findById(departmentId)
      if (department?.head) {
        session.assignedTo = department.head
      }
    } else {
      const technician = await User.findOne({
        role: 'technician',
        organization: session.organization,
      })
      if (technician) {
        session.assignedTo = technician._id
      }
    }

    await session.save()

    // Create system message
    const systemMessage = await ChatMessage.create({
      session: session._id,
      sender: 'system',
      content: 'Your conversation has been escalated to a technician. They will respond shortly.',
      messageType: 'system',
    })

    res.json({
      session,
      message: systemMessage,
    })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

export default router

File: backend/app/api/v1/endpoints/chatbot.py
"""
Chatbot endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from typing import List, Optional
from app.middleware.auth import get_current_user
from app.db.database import get_database
from bson import ObjectId
from datetime import datetime
import secrets

router = APIRouter()


@router.post("/session")
async def create_session(session_data: dict, current_user: dict = Depends(get_current_user)):
    """Create chat session"""
    db = await get_database()
    
    session_doc = {
        "user": ObjectId(current_user["id"]),
        "platform": session_data.get("platform", "web"),
        "created_at": datetime.utcnow(),
        "messages": []
    }
    
    result = await db.chat_sessions.insert_one(session_doc)
    session_doc["id"] = str(result.inserted_id)
    session_doc["user"] = str(session_doc["user"])
    del session_doc["_id"]
    
    return {"session": session_doc, "messages": []}


@router.post("/message")
async def send_message(
    message: Optional[str] = None,
    session_id: Optional[str] = None,
    attachments: List[UploadFile] = File([]),
    current_user: dict = Depends(get_current_user)
):
    """Send message to chatbot"""
    db = await get_database()
    
    # Get or create session
    if not session_id:
        session_doc = {
            "user": ObjectId(current_user["id"]),
            "platform": "web",
            "created_at": datetime.utcnow(),
            "messages": []
        }
        session_result = await db.chat_sessions.insert_one(session_doc)
        session_id = str(session_result.inserted_id)
    
    # Create message
    message_doc = {
        "message": message or "",
        "user": current_user["id"],
        "user_name": current_user["name"],
        "attachments": [],
        "created_at": datetime.utcnow(),
        "is_bot": False
    }
    
    # Handle attachments
    # TODO: Save attachments to disk
    
    # Add message to session
    await db.chat_sessions.update_one(
        {"_id": ObjectId(session_id)},
        {"$push": {"messages": message_doc}}
    )
    
    # TODO: Process message with chatbot and generate response
    bot_response = {
        "message": "This is a placeholder response. Chatbot implementation pending.",
        "created_at": datetime.utcnow(),
        "is_bot": True
    }
    
    await db.chat_sessions.update_one(
        {"_id": ObjectId(session_id)},
        {"$push": {"messages": bot_response}}
    )
    
    return {
        "message": bot_response,
        "session_id": session_id
    }


@router.post("/create-ticket")
async def create_ticket_from_chat(ticket_data: dict, current_user: dict = Depends(get_current_user)):
    """Create ticket from chat session"""
    db = await get_database()
    
    # Generate ticket ID
    import random
    import string
    ticket_id = f"TKT-{''.join(random.choices(string.ascii_uppercase + string.digits, k=8))}"
    
    ticket_doc = {
        "ticket_id": ticket_id,
        "title": ticket_data.get("title", "Ticket from Chat"),
        "description": ticket_data.get("description", ""),
        "status": "open",
        "priority": ticket_data.get("priority", "medium"),
        "creator": ObjectId(current_user["id"]),
        "attachments": [],
        "comments": [],
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }
    
    if ticket_data.get("category"):
        ticket_doc["category"] = ObjectId(ticket_data["category"])
    if ticket_data.get("department"):
        ticket_doc["department"] = ObjectId(ticket_data["department"])
    if current_user.get("organization"):
        ticket_doc["organization"] = ObjectId(current_user["organization"])
    
    result = await db.tickets.insert_one(ticket_doc)
    
    return {
        "id": str(result.inserted_id),
        "ticket_id": ticket_id,
        "message": "Ticket created successfully"
    }


@router.get("/history")
async def get_chat_history(
    userId: Optional[str] = None,
    limit: int = 50,
    current_user: dict = Depends(get_current_user)
):
    """Get chat history"""
    db = await get_database()
    query = {}
    
    if userId and current_user.get("role") == "admin":
        query["user"] = ObjectId(userId)
    else:
        query["user"] = ObjectId(current_user["id"])
    
    cursor = db.chat_sessions.find(query).sort("created_at", -1).limit(limit)
    sessions = await cursor.to_list(length=limit)
    
    result = []
    for session in sessions:
        session["id"] = str(session["_id"])
        session["user"] = str(session["user"])
        del session["_id"]
        result.append(session)
    
    return result


@router.get("/session/{session_id}")
async def get_session(session_id: str, current_user: dict = Depends(get_current_user)):
    """Get chat session by ID"""
    db = await get_database()
    
    try:
        session = await db.chat_sessions.find_one({"_id": ObjectId(session_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found"
        )
    
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found"
        )
    
    # Check permissions
    if str(session["user"]) != current_user["id"] and current_user.get("role") != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized"
        )
    
    session["id"] = str(session["_id"])
    session["user"] = str(session["user"])
    del session["_id"]
    
    return session


@router.post("/escalate")
async def escalate_chat(escalate_data: dict, current_user: dict = Depends(get_current_user)):
    """Escalate chat to human agent"""
    db = await get_database()
    
    session_id = escalate_data.get("sessionId")
    department_id = escalate_data.get("departmentId")
    
    await db.chat_sessions.update_one(
        {"_id": ObjectId(session_id)},
        {
            "$set": {
                "escalated": True,
                "escalated_to": ObjectId(department_id) if department_id else None,
                "escalated_at": datetime.utcnow()
            }
        }
    )
    
    return {"message": "Chat escalated successfully"}
